src/classes/BDI_DataImportService.cls

```diff
@@ -1725,68 +1723,141 @@ global with sharing class BDI_DataImportService {
+    * @description determine matching status based on campaign ID and/or name
+    * @param dataImport reference to the Data Import object to check and update
+    * @return void
+    */
+    private void setCampaignImportStatusField(DataImport__c dataImport) {
+
+        String campaignName = dataImport.Donation_Campaign_Name__c;
+        String currentImportStatus = dataImport.DonationCampaignImportStatus__c; 
+        Id campaignId = dataImport.DonationCampaignImported__c;
+
+        // if we have a campaign Id, and it isn't populated from previous dry-run match, set the status
+        if (campaignId != null && currentImportStatus != System.Label.bdiDryRunMatched) {
+            dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatchedId;
+        } else if (campaignName != null) {
+            Boolean campaignNameFound = campaignNameToCampaignId.containsKey(campaignName);
+            if (campaignNameFound) {
+                // The campaign exists!
+                dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatched;
+            } else {
+                dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatchedNone;                
             }
         }
+    }
 
-        // add the new campaigns to our map
-        for (Campaign cmp : listCmpNew) {
-            mapStrNameToCampaignId.put(cmp.Name, cmp.Id);
+    /*******************************************************************************************************
+     * @description If the data import needs a match and the name matches a current campaign, populate the lookup
+     * with the matched campaign id
+     * @param dataImport reference to the Data Import object to check and update
+     * @param campaignName the name of the Campaign in the Donation_Campaign_Name__c field on the Data Import
+     * @return void
+     */
+    private void setCampaignIdIfMatchedByName(DataImport__c dataImport, String campaignName) {
+        if (dataImport.DonationCampaignImported__c == null && campaignNameToCampaignId.containsKey(campaignName)) {
+            dataImport.DonationCampaignImported__c = campaignNameToCampaignId.get(campaignName);
         }
+    }
 
+    /*******************************************************************************************************
+     * @description Track duplicate campaign names so we only insert 1 campaign of the same name, but we associate
+     * that campaign with all the DIs that reference it
+     * @param dataImport the Data Import object to check
+     * @param campaignName name of the Campaign in the Donation_Campaign_Name__c field on the Data Import
+     * @param newCampaignNameToDataImports reference to the running list of campaign names and the Data Imports that reference them
+     * @return void
+     */
+    private void associateCampaignWithDataImport(DataImport__c dataImport, String campaignName, Map<String, List<DataImport__c>> newCampaignNameToDataImports) {
```

Comment:
I'm inclined to make campaignNameToDataImports an instance variable instead of passing it in. What are the advantages of making it a parameter instead?

Reply:
Yeah, I had a similar feeling. Mostly to avoid cluttering the class, but I could try moving it!